text/3525-struct-target-feature.md

```diff
@@ -0,0 +1,280 @@
+# Summary
+
+[summary]: #summary
+
+Allow adding `#[target_feature(enable = "...")]` attributes to unit structs, and
+enable the corresponding target features to functions taking those structs as
+parameters.
+
+# Motivation
+
+[motivation]: #motivation
+
+Currently, the only way to tell the compiler it can assume the availability of
+hardware features is by annotating a function with the corresponding
+`#[target_feature]` attribute. This requires that the annotated function be
+marked as unsafe as the caller must check whether the features are available at
+runtime.  
+This also makes it difficult for library authors to use in certain situations, as
+they may not know which features the library user wants to detect, and at what
+level the dynamic dispatch should be done.
+
+Assume we want to implement a library function that multiplies a slice of `f64`
+values by `2.0`.
+
+```rust
+pub fn times_two(v: &mut [f64]) {
+    for v in v {
+        *v *= 2.0;
+    }
+}
+```
+
+Generally speaking, during code generation, the compiler will only assume the
+availability of globally enabled target features (e.g., `sse2` on `x86-64`
+unless additional feature flags are passed to the compiler).
+
+This means that if the code is run on a machine with more efficient features
+such as `avx2`, the function will not be able to make good use of them.
+
+To improve performance, the library author may decide to add runtime feature
+detection to their implementation, choosing subsets of features to detect.
+
+```rust
+#[inline(always)]
+fn times_two_generic(v: &mut [f64]) {
+    for v in v {
+        *v *= 2.0;
+    }
+}
+
+#[target_feature(enable = "avx")]
+unsafe fn times_two_avx(v: &mut [f64]) {
+    times_two_generic(v);
+}
+
+#[target_feature(enable = "avx512f")]
+unsafe fn times_two_avx512f(v: &mut [f64]) {
+    times_two_generic(v);
+}
+
+pub fn times_two(v: &mut[f64]) {
+    if is_x86_feature_detected!("avx512f") {
+        times_two_avx512f(v);
+    } else if is_x86_feature_detected!("avx") {
+        times_two_avx(v);
+    } else {
+        times_two_generic(v);
+    }
+}
+```
+
+This decision, however, comes with a few drawbacks:
+
+- The runtime dispatch now implies that the code has some additional overhead
+  to detect the hardware features, which can harm performance for small
+  slices.
+- The addition of more code paths increases binary size.
+- The dispatch acts as a barrier that prevents inlining, which can prevent
+  compiler optimizations at the call-site.
+- This requires adding unsafe code to the library, which has a maintenance cost.
+
+The proposed alternative offers solutions for these issues.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Suppose we define the following structs:
+
+```rust
+#[target_feature(enable = "avx")]
+#[derive(Clone, Copy, Debug)]
+pub struct Avx;
```

Comment:
I get that this must have language support to work, but it is not clear to me why this must be a language feature exposed to users? Maybe the standard library should be the one defining the interface to these structs, and hide the language feature that makes it work?

Reply:
I had the same thought. When I asked on the Zulip thread for this pre-RFC/RFC I was told that such a design would essentially be fine **if** the target feature standard library structs could still compose into larger structs:
```rust
struct AvxFma(Avx, Fma);
```

Reply:
right, the composability idea was mentioned in the "future possibilities" of the RFC. i didn't know if it would be better to prioritize that or visibility to users